rlclientlib/federation/vw_trainable_model.cc

```diff
@@ -0,0 +1,441 @@
+#include "federation/vw_trainable_model.h"
+
+#include "constants.h"
+#include "err_constants.h"
+#include "joiners/example_joiner.h"
+#include "joiners/multistep_example_joiner.h"
+#include "parse_example_binary.h"
+#include "str_util.h"
+#include "vw/config/options_cli.h"
+#include "vw/core/learner.h"
+#include "vw/core/parse_primitives.h"
+#include "vw/core/shared_data.h"
+#include "vw/core/vw.h"
+#include "vw/io/logger.h"
+
+namespace
+{
+// Helper function to train model on VW::multi_ex
+// examples is cleared at the end of this function
+// returns number of examples learned
+int learn_and_finish_examples(VW::workspace& vw, VW::multi_ex& examples)
+{
+  if (examples.empty()) { return 0; }
+
+  VW::setup_examples(vw, examples);
+
+  if (vw.l->is_multiline())
+  {
+    vw.learn(examples);
+    vw.finish_example(examples);
+    examples.clear();
+    return 1;
+  }
+
+  // single line
+  for (auto example : examples) { vw.learn(*example); }
+  for (auto example : examples) { vw.finish_example(*example); }
+  int size = examples.size();
+  examples.clear();
+  return size;
+}
+
+// Helper function to call finish_example on VW::multi_ex
+// examples is cleared at the end of this function
+void finish_examples(VW::workspace& vw, VW::multi_ex& examples)
+{
+  if (examples.empty()) { return; }
+
+  if (vw.l->is_multiline()) { vw.finish_example(examples); }
+  else
+  {
+    for (auto example : examples) { vw.finish_example(*example); }
+  }
+
+  examples.clear();
+}
+
+void vw_log_to_trace_logger(void* trace_logger, VW::io::log_level log_level, const std::string& msg)
+{
+  if (trace_logger == nullptr) { return; }
+  auto i_trace_ptr = static_cast<reinforcement_learning::i_trace*>(trace_logger);
+  switch (log_level)
+  {
+    case VW::io::log_level::TRACE_LEVEL:
+    case VW::io::log_level::DEBUG_LEVEL:
+      TRACE_DEBUG(i_trace_ptr, msg);
+      break;
+
+    case VW::io::log_level::INFO_LEVEL:
+      TRACE_INFO(i_trace_ptr, msg);
+      break;
+
+    case VW::io::log_level::WARN_LEVEL:
+      TRACE_WARN(i_trace_ptr, msg);
+      break;
+
+    case VW::io::log_level::ERROR_LEVEL:
+    case VW::io::log_level::CRITICAL_LEVEL:
+      TRACE_ERROR(i_trace_ptr, msg);
+      break;
+
+    default:
+      break;
+  }
+}
+}  // namespace
+
+namespace reinforcement_learning
+{
+int trainable_vw_model::create(std::unique_ptr<trainable_vw_model>& output, const utility::configuration& config,
+    i_trace* trace_logger, api_status* status)
+{
+  int protocol_version = config.get_int(name::PROTOCOL_VERSION, 0);
```

Comment:
kind of nit since it is not affecting final result here, but in other places default version is 1, so it is maybe better to be consistent about it.